subprojects/core-api/src/main/java/org/gradle/api/artifacts/PublishArtifactSet.java

```diff
@@ -15,13 +15,28 @@
  */
 package org.gradle.api.artifacts;
 
+import org.gradle.api.Action;
 import org.gradle.api.Buildable;
 import org.gradle.api.DomainObjectSet;
+import org.gradle.api.Incubating;
 import org.gradle.api.file.FileCollection;
+import org.gradle.api.file.FileSystemLocation;
+import org.gradle.api.provider.Provider;
 
 /**
  * A set of artifacts to be published.
  */
 public interface PublishArtifactSet extends DomainObjectSet<PublishArtifact>, Buildable {
+    /**
+     * Registers a file for publication.
+     *
+     * @param artifactFile provider to file to register
+     * @param configuration action to configure a publish artifact
+     * @return a Provider to the publish artifact
+     * @since 5.1
+     */
+    @Incubating
+    Provider<PublishArtifact> register(Provider<? extends FileSystemLocation> artifactFile, Action<ConfigurablePublishArtifact> configuration);
```

Comment:
This should return a `DomainObjectProvider<ConfigurablePublishArtifact>` I think. We want the elements of all containers to be treated the same way.